### PR TITLE
Add port FEC histogram counter support.

### PR DESCRIPTION
**Pull request by mikeberesford**
Signed-off-by: Mike Beresford <mberes@google.com>

### CODE REVIEW

**Comment by kcudnik** [comment]
please fix errors
```
WARNING: star should be next to param name saiswitch.h:2783:     * SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_HISTOGRAM_S* attributes.
Running Aspell
Words to check: 3002
WARNING: Word 'codeword' is misspelled saiport.h 2761:     * @brief Port FEC codeword symbol error counters.
WARNING: Word 'codewords' is misspelled saiport.h 2774:    /** Count of FEC codewords with 1 symbol error. */
WARNING: Word 'retreived' is misspelled saiport.h 2769:     * of errors that can be detected may be retreived via the
```

**Comment by mikeberesford** [comment]
Meta checker errors fixed

**Comment by rlhui** [comment]
@prgeor , please help review? thx.

**Review by kcudnik — APPROVED** [review]

**Comment by mikeberesford** [comment]
@JaiOCP please review as well

**Comment by lguohan** [comment]
can we have code example how these attributes are used?

**Comment by mikeberesford** [comment]
@lguohan is there anything in particular you'd like to see? The SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_HISTOGRAM_* are accessed in the same way as all the rest of the port counters. SAI_PORT_ATTR_MAX_FEC_ERRORS_DETECTABLE could be used to determine the number of counters to collect if desired.

File: inc/saiport.h [hunk 1]
```diff
@@ -2749,6 +2757,68 @@ typedef enum _sai_port_stat_t
...
+    SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_S12,
+
+    /** Count of FEC codewords with 13 symbol errors. */
+    SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_S13,
+
+    /** Count of FEC codewords with 14 symbol errors. */
+    SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_S14,
+
+    /** Count of FEC codewords with 15 symbol errors. */
+    SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_S15,
```
Comment (prgeor) [inline inc/saiport.h:2820]: what will be the value of SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_SXX if XX is > max detectable error? 0 ?

**Review by JaiOCP — APPROVED** [review]

File: inc/saiport.h [hunk 2]
```diff
@@ -2151,6 +2151,14 @@ typedef enum _sai_port_attr_t
      */
     SAI_PORT_ATTR_FABRIC_ISOLATE,
 
+    /**
+     * @brief Query the maximum number of errors the current FEC can detect.
```
[the remark this answers is not in the document]
Comment (mikeberesford) [reply inc/saiport.h]: Good suggestion, updated

File: inc/saiport.h (hunk 1 above)
Comment (mikeberesford) [reply inc/saiport.h:2827]: I've updated the documentation to indicate that if it's greater than max detectable the next bin is used and added an additional counter S16. Let me know what you think.

File: inc/saiport.h [hunk 3]
```diff
@@ -2749,6 +2757,68 @@ typedef enum _sai_port_stat_t
...
+
+    /** Count of FEC codewords with 1 symbol error. */
+    SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_S1,
+
+    /** Count of FEC codewords with 2 symbol errors. */
+    SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_S2,
+
+    /** Count of FEC codewords with 3 symbol errors. */
+    SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_S3,
+
```
[the remark this answers is not in the document]
Comment (mikeberesford) [reply inc/saiport.h]: These are just finer-grained statistics. They help us see if the links are experiencing bursty errors or other time-correlated problems which are driving the link toward uncorrectable errors.

File: inc/saiport.h [hunk 4]
```diff
@@ -2749,6 +2758,77 @@ typedef enum _sai_port_stat_t
...
+     * of FEC in use. For instance, RS-528 FEC supports detection of up to 7
+     * symbol errors, while RS-544 FEC supports detection of up to 15 symbol
+     * errors. The maximum number of errors that can be detected by the port's
+     * current FEC mode may be retrieved via the
+     * SAI_PORT_ATTR_MAX_FEC_SYMBOL_ERRORS_DETECTABLE port attribute. If the
+     * codeword contains more than SAI_PORT_ATTR_MAX_FEC_SYMBOL_ERRORS_DETECTABLE,
+     * the errors are placed in the next higher counter (so for the examples
+     * above, SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_S8 would be used for
+     * greater than 7 symbol errors when RS-528 FEC is used, and
+     * SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_S16 for greater than 16 symbol
```
Comment (rlhui) [inline inc/saiport.h:2777]: S16 is not greater than 16.

**Comment by rlhui** [comment]
> @lguohan is there anything in particular you'd like to see? The SAI_PORT_STAT_IF_IN_FEC_CODEWORD_ERRORS_HISTOGRAM_* are accessed in the same way as all the rest of the port counters. SAI_PORT_ATTR_MAX_FEC_ERRORS_DETECTABLE could be used to determine the number of counters to collect if desired.

example code of how a NOS should read and interpret it.

File: inc/saiport.h (hunk 1 above)
Comment (prgeor) [reply inc/saiport.h:2827]: ok

**Review by prgeor — APPROVED** [review]

File: inc/saiport.h (hunk 4 above)
Comment (mikeberesford) [reply inc/saiport.h]: Thanks, fixed.

**Pull request merged**